File: fixtures/scale/java/large/src/cryptobench/ciphers/crc32/Crc32Hash.java
package cryptobench.ciphers.crc32;

import cryptobench.core.HashFunction;
import cryptobench.util.Bytes;

/** Bitwise CRC-32 with the reflected polynomial, no lookup table. */
public class Crc32Hash implements HashFunction {
    @Override
    public String name() {
        return "crc32";
    }

    @Override
    public long digest(String input) {
        long state = 0xFFFFFFFFL;
        for (byte b : Bytes.of(input)) {
            state ^= (b & 0xFF);
            for (int bit = 0; bit < 8; bit++) {
                state = (state >>> 1) ^ (0xEDB88320L & -(state & 1));
            }
        }
        return state ^ 0xFFFFFFFFL;
    }
}

File: fixtures/scale/dart/large/lib/ciphers/crc32/crc32_hash.dart
import '../../core/hash_function.dart';
import '../../util/bytes.dart';

/// Bitwise CRC-32 with the reflected polynomial, no lookup table.
class Crc32Hash implements HashFunction {
  @override
  String name() => 'crc32';

  @override
  int digest(String input) {
    var state = 0xFFFFFFFF;
    for (final b in bytesOf(input)) {
      state = (state ^ b) & 0xFFFFFFFF;
      for (var bit = 0; bit < 8; bit++) {
        final mask = -(state & 1) & 0xFFFFFFFF;
        state = ((state >> 1) ^ (0xEDB88320 & mask)) & 0xFFFFFFFF;
      }
    }
    return state ^ 0xFFFFFFFF;
  }
}
